firmware/application/ui/ui_freqlist.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2015 Jared Boone, ShareBrained Technology, Inc.
+ *
+ * This file is part of PortaPack.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "ui_freqlist.hpp"
+
+#include "baseband_api.hpp"
+#include "utility.hpp"
+
+#include <algorithm>
+
+namespace ui {
+FreqManUIList::FreqManUIList(
+    Rect parent_rect,
+    bool instant_exec)
+    : Widget{parent_rect},
+      instant_exec_{instant_exec} {
+    this->set_focusable(true);
+}
+
+uint8_t FreqManUIList::set_highlighted_index(int index) {
+    if ((unsigned)(current_index + index) >= freqlist_db.size())
+        return highlighted_index;
+    if (freqlist_db.size() == 0) {
+        current_index = highlighted_index = 0;
+        return 0;
+    }
+    if (index < 0) {
+        index = 0;
+        if (current_index > 0)
+            current_index--;
+    }
+    if (index >= freqlist_nb_lines) {
+        index = freqlist_nb_lines - 1;
+        if ((unsigned)(current_index + index) < freqlist_db.size())
```

Comment:
I wonder if you want size() - 1 here. If current_index == size()-1 and index is 0, won't the next line end up with an invalid index (== size())

Reply:
It should ne show up if it's empty, because on top of that current_index + index would forcibly >= db.size